pkg/distributor/distributor.go

```diff
@@ -2281,66 +2284,107 @@ func (d *Distributor) MetricsMetadata(ctx context.Context, req *ingester_client.
 
 // UserStats returns statistics about the current user.
 func (d *Distributor) UserStats(ctx context.Context, countMethod cardinality.CountMethod) (*UserStats, error) {
-	//nolint:staticcheck
-	replicationSet, err := d.getIngesterReplicationSetForQuery(ctx)
+	replicationSets, err := d.getIngesterReplicationSetsForQuery(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	// If we have a single zone, we can't tolerate any errors.
-	if replicationSet.ZoneCount() == 1 {
-		replicationSet.MaxErrors = 0
-	}
-
 	type zonedUserStatsResponse struct {
 		zone string
 		resp *ingester_client.UserStatsResponse
 	}
 
+	// When ingest storage is disabled, if ingesters are running in a single zone we can't tolerate any errors.
+	// In this case we expect exactly 1 replication set.
+	if !d.cfg.IngestStorageConfig.Enabled && len(replicationSets) == 1 && replicationSets[0].ZoneCount() == 1 {
+		replicationSets[0].MaxErrors = 0
+	}
+
 	ingesterCountMethod, err := toIngesterCountMethod(countMethod)
 	if err != nil {
 		return nil, err
 	}
-	req := &ingester_client.UserStatsRequest{
-		CountMethod: ingesterCountMethod,
-	}
-	resps, err := ring.DoUntilQuorum[zonedUserStatsResponse](ctx, replicationSet, d.queryQuorumConfig(ctx, replicationSet), func(ctx context.Context, desc *ring.InstanceDesc) (zonedUserStatsResponse, error) {
-		poolClient, err := d.ingesterPool.GetClientForInstance(*desc)
-		if err != nil {
-			return zonedUserStatsResponse{}, err
-		}
 
-		client := poolClient.(ingester_client.IngesterClient)
-		resp, err := client.UserStats(ctx, req)
+	var (
+		req          = &ingester_client.UserStatsRequest{CountMethod: ingesterCountMethod}
+		quorumConfig = d.queryQuorumConfigForReplicationSets(ctx, replicationSets)
+
+		responsesByReplicationSetMx = sync.Mutex{}
+		responsesByReplicationSet   = make(map[int][]zonedUserStatsResponse, len(replicationSets))
```

Comment:
We could index replication set by index, but we still need a mutex to append to `[]zonedUserStatsResponse` (and index ingesters too is pretty annoying). I don't see a big push to avoid a mutex which locks a very fast operation.

Reply:
Good suggestion. Done in https://github.com/grafana/mimir/pull/7402/commits/8c721dfdc890bdc0e2368fc10a6b0184977e67af.